File: udgaard/src/main/java/com/skrymer/udgaard/integration/ovtlyr/DataLoader.java
package com.skrymer.udgaard.integration.ovtlyr;

import java.util.Arrays;

import org.springframework.stereotype.Component;

import com.skrymer.udgaard.model.MarketSymbol;
import com.skrymer.udgaard.repository.MarketBreadthRepository;

/**
 * Loads data from ovtlyr into mongo.
 */
@Component
public class DataLoader {

    private final MarketBreadthRepository marketBreadthRepository;
    private final OvtlyrClient ovtlyrClient;

    public DataLoader(OvtlyrClient ovtlyrClient, MarketBreadthRepository marketBreadthRepository){
        this.ovtlyrClient = ovtlyrClient;
        this.marketBreadthRepository = marketBreadthRepository;
    }

    public void loadData(){
        try {
            loadMarkBreadthForAllSectors();            
        } catch (Exception e) {
            // TODO: handle exception logging
        }
    }

    private void loadMarkBreadthForAllSectors(){
        Arrays.asList(MarketSymbol.values()).forEach(symbol -> {
			var response = ovtlyrClient.getMarketBreadth(((MarketSymbol)symbol).name());
			marketBreadthRepository.save(response.toModel());
			System.out.println(response);
		});
    }
}

File: udgaard/src/main/kotlin/com/skrymer/udgaard/integration/ovtlyr/DataLoader.kt
package com.skrymer.udgaard.integration.ovtlyr

import com.skrymer.udgaard.model.EtfEntity
import com.skrymer.udgaard.model.EtfSymbol
import com.skrymer.udgaard.model.SectorSymbol
import com.skrymer.udgaard.model.Stock
import com.skrymer.udgaard.model.StockSymbol
import com.skrymer.udgaard.service.BreadthService
import com.skrymer.udgaard.service.EtfService
import com.skrymer.udgaard.service.StockService
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.async
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.supervisorScope
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

/**
 * Loads data from ovtlyr into mongo.
 */
@Component
class DataLoader(
  private val breadthService: BreadthService,
  private val stockService: StockService,
  private val etfService: EtfService,
) {
  fun loadData() {
    runBlocking { loadBreadthForAll() }
    runBlocking { loadStocks(true) }
    runBlocking { loadEtfs() }
  }

  fun loadStocks(forceFetch: Boolean = false): List<Stock> =
    stockService.getStocksBySymbols(
      StockSymbol.entries.map {
        it.symbol
      },
      forceFetch,
    )

  fun loadEtfs(): List<EtfEntity> {
    val logger = LoggerFactory.getLogger("EtfLoader")
    logger.info("Loading ${EtfSymbol.entries.size} ETFs from Ovtlyr")

    return EtfSymbol.entries.mapNotNull { etfSymbol ->
      runCatching {
        etfService.refreshEtf(etfSymbol.name)
      }.onFailure { e ->
        logger.warn("Failed to fetch ETF ${etfSymbol.name}: ${e.message}", e)
      }.getOrNull()
    }
  }

  @OptIn(ExperimentalCoroutinesApi::class)
  private suspend fun loadBreadthForAll() =
    supervisorScope {
      val logger = LoggerFactory.getLogger("BreadthLoader")
      val limited = Dispatchers.IO.limitedParallelism(10)

      // Load market breadth (FULLSTOCK)
      async(limited) {
        runCatching { breadthService.getMarketBreadth(refresh = true) }
          .onFailure { e -> logger.warn("Failed to fetch market breadth: {}", e.message, e) }
      }

      // Load all sector breadth
      SectorSymbol.entries.forEach { sector ->
        async(limited) {
          runCatching { breadthService.getSectorBreadth(sector, refresh = true) }
            .onFailure { e -> logger.warn("Failed to fetch sector={}: {}", sector, e.message, e) }
        }
      }
    }
}
